File: javascript/src/problem12.js
"use strict";

function problem() {

    let number =1;

    let counter=2;

    while(divisors(number)<=500){
        number += counter;
        counter++;
    }

    return number;

    function divisors(number){
        let divisorsNumber=2;
        const square=Math.sqrt(number);
        for (let i=2;i<=square;i++){
            if (number % i==0){
                divisorsNumber+=2;
            }
        }

        if (square*square== number){
            divisorsNumber--;
        }
        return divisorsNumber;
    }
}

module.exports = problem;

File: python/main/problem12.py
import math


def get_problem12():
    return find_first_triangle_number_with_divisors(500)


def number_of_divisors(triangle_number):
    square = int(math.sqrt(triangle_number))

    counter = 2
    for i in range(2, square):
        if triangle_number % i == 0:
            counter += 2

    if square * square == triangle_number:
        counter += 1

    return counter

    # return using_primes(triangle_number)


# def using_primes(triangle_number):
#     return reduce(mul, [item[1] + 1 for item in find_prime_decomposition(triangle_number)])


def find_first_triangle_number_with_divisors(min_number_of_divisors):
    triangle_number = 28
    next = 8

    while number_of_divisors(triangle_number) <= min_number_of_divisors:
        triangle_number += next
        next += 1

    return triangle_number
